zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,182 @@
+   `push_account_id`, and `encrypted_push_registration`.
+
+   The push bouncer:
+   1. Decrypts `push_registration`.
+
+   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
+      A coded `JsonableError` should be used if the timestamp is stale.
+
+   1. Checks that it doesn’t already have a `RemotePushDevice` for this `token`
+      and `push_account_id`. This check helps to avoid creating a duplicate
+      record if the client retries when a previous attempt actually succeeded.
+
+   1. Creates a `RemotePushDevice` instance, generating a `device_id`.
+
+   **On success**, the bouncer replies to the server with `device_id`.
+   The server stores `device_id` in the `PushDevice`.
+    
+   **On error**, the server keep retrying for a long time. Like maybe
+   after a day we’d pass the error to the server admin and the user
+   (because a day is longer than any minor outage), but we’d keep retrying for a week.
+    
+   Then finally after a long time perhaps we give up retrying and forget
+   that `push_account_id`. That causes the client to retry from scratch,
+   if it’s still in use. That way if the client keeps being used, we keep
+   retrying and keep sending the server admin error messages about it
+   — appropriately so — but we don’t accumulate ghost clients.
+
+### To send a notification
+
+1. The server encrypts the notification content to `push_key` as `encrypted_content`.
+
+1. The server makes a batched API call to the push bouncer (authenticated
+   with the server’s API key) with arguments `device_id` and `encrypted_content`.
+   - (That is, the HTTP request has a list of `(device_id, encrypted_content)` pairs.)
+   - (It’s OK, just inefficient, if all batches generated by the server are,
+     in practice, 1 notice in an initial version)
+
+1. The push bouncer sends a push notification to the client via APNs or FCM
+   using `token_kind` and `token`, with a payload including `push_account_id`
+   and `encrypted_content`.
+
+1. The client uses `push_account_id` to look up the right `push_key`,
+   decrypts `encrypted_content` with it, and ends up with a set of
+   notifications to process.
+
+1. The client sorts the notifications by `message_id`, with `remove`
+   notifications cancelling `add` notifications, and passes them into
+   its display layer.
+
+### Miscellaneous
+
+- If APNs or FCM indicates to the push bouncer that the token is invalid
+  or expired, the push bouncer sets `expired_time` to now. In that case
+  — or if the push bouncer gets a notification for an invalid device or
+  a device with `expired_time` already set — it returns an error to the server
+  that the `device_id` is no longer valid, and the server removes it.
+  
+- The server gives the client a list of the user’s currently valid `push_account_id`s
+  as part of `register` response, and keeps it updated through events.
+  - This allows the client to determine when it’s necessary to reregister the push device.
```

Comment:
Copying over a comment from the old doc:
> I think the idea here (which should be spelled out) is that if the client doesn’t see on the list the `push_account_id` that it has for this account, then it knows it needs to reregister. The consequence of that is that the list should include all the `push_account_id`s that the client doesn’t need to act on by reregistering, which in particular means it should include any that the server is still working on setting up with the bouncer.
>
> [Then in addition to that information, i]ncluding the status (i.e. “has successfully registered with bouncer”, i.e. “has a non-null `device_id`") may be useful for a troubleshooting UI.

Reply:
Rewrote this section. Give it a read.